optimum/habana/transformers/models/mistral/modeling_mistral.py

```diff
@@ -23,330 +23,537 @@
 import warnings
 from typing import List, Optional, Tuple, Union
 
+import habana_frameworks.torch.core as htcore
 import torch
 from torch import nn
 from torch.nn import CrossEntropyLoss
 from transformers.cache_utils import Cache, DynamicCache
 from transformers.modeling_attn_mask_utils import _prepare_4d_causal_attention_mask_for_sdpa
 from transformers.modeling_outputs import BaseModelOutputWithPast, CausalLMOutputWithPast
-from transformers.models.mistral.modeling_mistral import MistralForCausalLM, apply_rotary_pos_emb, repeat_kv
+from transformers.models.mistral.configuration_mistral import MistralConfig
+from transformers.models.mistral.modeling_mistral import (
+    MistralAttention,
+    MistralDecoderLayer,
+    MistralForCausalLM,
+    MistralMLP,
+    MistralModel,
+    MistralRMSNorm,
+    apply_rotary_pos_emb,
+)
 from transformers.utils import logging
 
 from ...modeling_attn_mask_utils import (
     _gaudi_prepare_4d_causal_attention_mask,
 )
 
 
+try:
+    from habana_frameworks.torch.hpex.normalization import FusedRMSNorm as FusedRMSNorm
+except ImportError:
+    print("Not using HPU fused kernel for RMSNorm")
+    FusedRMSNorm = None
+
 logger = logging.get_logger(__name__)
 
 
-def gaudi_mistral_attention_forward(
-    self,
-    hidden_states: torch.Tensor,
-    attention_mask: Optional[torch.Tensor] = None,
-    position_ids: Optional[torch.LongTensor] = None,
-    past_key_value: Optional[Cache] = None,
-    output_attentions: bool = False,
-    use_cache: bool = False,
-    token_idx: Optional[torch.Tensor] = None,
-    **kwargs,
-) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+def update(prev, cur, dim, idx):
+    orig_cur = cur
+    if prev.shape == cur.shape:
+        # Initialize
+        prev.copy_(cur)
+        return orig_cur
+    assert cur.shape[2] == 1, f"Cannot update kv-cache. Unsupported shapes. prev:{prev.shape} cur:{cur.shape}"
+    if idx is not None:
+        return prev.index_copy_(dim, idx - 1, cur)
+    else:
+        return torch.cat((prev, cur), dim=dim)
+
+
+# Copied from transformers.models.llama.modeling_llama.repeat_kv
```

Comment:
To remove

Reply:
this is in the spirit of backtracking to source. we have a lot of these "copied from" lines:

https://github.com/search?q=repo%3Ahuggingface%2Foptimum-habana+%22Copied+from%22&type=code

I can remove it if you still want to

Reply:
I know, I introduced these comments :grin: 
But this one is from Transformers: https://github.com/huggingface/transformers/blob/66ce9593fdb8e340df546ddd0774eb444f17a12c/src/transformers/models/mistral/modeling_mistral.py#L182
Note how it refers to `modeling_llama` and not `modeling_mistral`, they use this kind of comments there to have some kind of automatic duplication so that it is automatically updated if `modeling_llama.repeat_kv` is modified. But we don't use this in Optimum Habana.
The similar reference we added is right below at the beginning of the docstring.